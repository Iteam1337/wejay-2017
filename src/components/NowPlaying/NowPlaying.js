import React from 'react'

const NowPlaying = () => {
  return <div />
}

export default NowPlaying
