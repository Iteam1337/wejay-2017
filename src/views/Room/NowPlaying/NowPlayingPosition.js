// @flow

import * as React from 'react'
import * as WejayApi from '__generated__/types.flow'
import styled from 'styled-components'
import TrackDuration from 'components/Track/TrackDuration'

type NowPlayingPositionProps = {
  isPlaying: boolean,
  track: WejayApi.TrackInfoFragment,
}

type NowPlayingPositionState = {
  barWidth: number,
  playedDuration: number,
}

const PositionWrap = styled.div`
  align-items: center;
  display: none;
  grid-column-gap: 10px;
  grid-template-columns: auto 1fr auto;

  @media (min-width: 769px) {
    display: grid;
  }
`

const PositionBarWrap = styled.div`
  height: 5px;
`

const PositionBar = styled.div`
  background: rgba(255, 134, 0, 0.4);
  border-radius: 0 2px 0 0;
  height: 5px;
  transition: width 100ms ease-in-out;
`

class NowPlayingPosition extends React.Component<
  NowPlayingPositionProps,
  NowPlayingPositionState
> {
  timer: IntervalID

  state = {
    barWidth: 0,
    playedDuration: 0,
  }

  componentDidMount () {
    if (this.props.isPlaying) {
      this.startTrackingDuration()
    }
  }

  componentWillUnmount () {
    clearInterval(this.timer)
  }

  componentWillReceiveProps (nextProps: NowPlayingPositionProps) {
    const { track } = this.props

    if (track.spotifyUri !== nextProps.track.spotifyUri) {
      this.startTrackingDuration()
    }

    if (nextProps.isPlaying === false) {
      clearInterval(this.timer)
    }

    if (nextProps.isPlaying === true) {
      this.startTrackingDuration()
    }
  }

  startTrackingDuration = () => {
    this.timer = setInterval(this.updateTimer, 1000)
  }

  updateTimer = () => {
    const { track } = this.props
    const started = track.started || 0

    this.setState(state => ({
      barWidth: state.playedDuration / track.duration * 100,
      playedDuration:
        state.playedDuration > 0
          ? state.playedDuration + 1000
          : Date.now() - started,
    }))
  }

  render () {
    const { track } = this.props
    const { barWidth, playedDuration } = this.state

    return (
      <PositionWrap>
        <TrackDuration duration={playedDuration} />
        <PositionBarWrap>
          <PositionBar
            style={{ width: `${barWidth < 100 ? barWidth : 100}%` }}
          />
        </PositionBarWrap>
        <TrackDuration duration={track.duration} />
      </PositionWrap>
    )
  }
}

export default NowPlayingPosition
