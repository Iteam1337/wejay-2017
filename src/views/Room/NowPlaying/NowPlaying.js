// @flow

import * as WejayApi from '__generated__/types.flow'
import React from 'react'
import Cover from 'components/Cover/Cover'
import Gravatar from 'components/Gravatar/Gravatar'
import styled from 'styled-components'
import TrackMeta from 'components/Track/TrackMeta'
import NowPlayingPosition from './NowPlayingPosition'

type NowPlayingProps = {
  isPlaying: boolean,
  track: ?WejayApi.TrackInfoFragment,
}

const NowPlayingWrap = styled.section`
  background-color: #fff;
  border-top: 1px solid #eaecef;
  bottom: 0;
  left: 0;
  padding: 20px;
  position: sticky;
  right: 0;
  z-index: 1;
`

const NowPlayingInner = styled.section`
  align-items: center;
  display: grid;
  grid-column-gap: 20px;
  grid-template-columns: auto 1fr 0 auto;
  margin-left: auto;
  margin-right: auto;
  max-width: 960px;

  @media (min-width: 769px) {
    grid-template-columns: auto 1fr 400px auto;
  }
`

const NowPlaying = ({ isPlaying, track }: NowPlayingProps) => {
  if (!track) {
    return null
  }

  return (
    <NowPlayingWrap>
      <NowPlayingInner>
        <Cover small track={track} width={40} />

        <TrackMeta artists={track.artists} name={track.name} />
        <NowPlayingPosition isPlaying={isPlaying} track={track} />

        {track.user && <Gravatar id={track.user.id} size={30} />}
      </NowPlayingInner>
    </NowPlayingWrap>
  )
}

export default NowPlaying
