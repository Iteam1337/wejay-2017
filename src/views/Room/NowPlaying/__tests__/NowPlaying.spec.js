import React from 'react'
import { shallow } from 'enzyme'
import NowPlaying from '../NowPlaying'

describe('components/NowPlaying', () => {
  let component

  beforeEach(() => {
    component = shallow(
      <NowPlaying
        track={{
          artists: [{ name: 'artist one' }, { name: 'artist two' }],
          user: {
            id: '1337',
          },
        }}
      />
    )
  })

  it('should return null if no track is provided', () => {
    component.setProps({
      track: null,
    })

    expect(component.type()).toBeNull()
  })

  it('renders NowPlaying', () => {
    expect(component).toMatchSnapshot()
  })
})
