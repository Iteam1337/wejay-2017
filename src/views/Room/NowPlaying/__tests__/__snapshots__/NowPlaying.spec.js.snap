// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/NowPlaying renders NowPlaying 1`] = `
<styled.section>
  <styled.section>
    <Component
      small={true}
      track={
        Object {
          "artists": Array [
            Object {
              "name": "artist one",
            },
            Object {
              "name": "artist two",
            },
          ],
          "user": Object {
            "id": "1337",
          },
        }
      }
      width={40}
    />
    <styled.div>
      <TrackArtist
        artists={
          Array [
            Object {
              "name": "artist one",
            },
            Object {
              "name": "artist two",
            },
          ]
        }
      />
      <styled.div />
    </styled.div>
    <PositionTracker
      track={
        Object {
          "artists": Array [
            Object {
              "name": "artist one",
            },
            Object {
              "name": "artist two",
            },
          ],
          "user": Object {
            "id": "1337",
          },
        }
      }
    />
    <Gravatar
      email=""
      id="1337"
      size={30}
    />
  </styled.section>
</styled.section>
`;
